seo4ajax.go

```diff
@@ -8,118 +8,171 @@ import (
+	Timeout time.Duration // retry timeout, defaults to 30s
+}
+
+// Client is the Seo4Ajax Client
 type Client struct {
-	APIHost         string
-	serverIP, token string
+	log     log.Logger
+	next    http.Handler
+	server  string
+	token   string
+	ip      string
+	timeout time.Duration
+	http    *http.Client
 }
 
-func New(serverIP, token string) (*Client, error) {
-	if serverIP == "" || net.ParseIP(serverIP) == nil {
-		return nil, errors.New("no ip address")
+// New creates a new Seo4Ajax client. Returns an error if no token is provided
+func New(cfg Config) (*Client, error) {
+	if cfg.Log == nil {
+		cfg.Log = log.NewNopLogger()
 	}
-
-	if token == "" {
+	if cfg.Server == "" {
+		cfg.Server = "http://api.seo4ajax.com"
+	}
+	if cfg.Token == "" {
 		return nil, ErrNoToken
 	}
+	if cfg.IP == "" {
+		cfg.IP = "127.0.0.1"
+	}
+	if cfg.Timeout < time.Second {
+		cfg.Timeout = 30 * time.Second
+	}
 
-	return &Client{
-		APIHost:  "http://api.seo4ajax.com",
-		serverIP: serverIP,
-		token:    token,
-	}, nil
+	c := &Client{
+		log:     cfg.Log,
+		server:  cfg.Server,
+		token:   cfg.Token,
+		ip:      cfg.IP,
+		timeout: cfg.Timeout,
+		next:    cfg.Next,
+	}
+	c.http = &http.Client{
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			return errRedirect
+		},
+	}
+	return c, nil
 }
 
 // IsPrerender returns true, when Seo4Ajax shall be used for the given http Request.
 // The logic is taken from https://github.com/seo4ajax/connect-s4a/blob/master/lib/connect-s4a.js
-func IsPrerender(req *http.Request) (usePrerender bool) {
-	if req.Method != "GET" && req.Method != "HEAD" {
+func IsPrerender(r *http.Request) bool {
+	if r.Method != "GET" && r.Method != "HEAD" {
 		return false
 	}
 
-	if strings.Contains(req.URL.RawQuery, "_escaped_fragment_") {
+	if strings.Contains(r.URL.RawQuery, "_escaped_fragment_") {
 		return true
 	}
 
-	if regexInvalidUserAgent.MatchString(req.Header.Get("User-Agent")) {
+	if regexInvalidUserAgent.MatchString(r.Header.Get("User-Agent")) {
 		return false
 	}
 
-	if regexPath.MatchString(req.URL.Path) {
+	if regexPath.MatchString(r.URL.Path) {
 		return false
 	}
 
-	return regexValidUserAgent.MatchString(req.Header.Get("User-Agent"))
+	return regexValidUserAgent.MatchString(r.Header.Get("User-Agent"))
 }
 
-// GetPrerenderedPage returns the prerendered html from the seo4ajax api. If no
-// token is given, it will return an error.
-func (c *Client) GetPrerenderedPage(w http.ResponseWriter, req *http.Request) (err error) {
-	var prerenderRequest *http.Request
-	prerenderRequest, err = http.NewRequest("GET", fmt.Sprintf("%s/%s%s", c.APIHost, c.token, cleanPath(req.URL)), nil)
-	if err != nil {
+// ServeHTTP will serve the prerendered page if this is a prerender request
+// or no upstream handler is set. Otherwise it will just invoke the upstream
+// handler
+func (c *Client) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if IsPrerender(r) || c.next == nil {
```

Comment:
wouldn't `c.next == nil` cause an infinite loop? When the prerendered site is requested, seo4ajax will request the normal site again, which will cause the same flow again